File: packages/go-sdk/authoring/run_context.go
// Package authoring provides the job and workflow definition DSL for the
// Strait Go SDK.
package authoring

import (
	"context"
	"log/slog"
)

// RunContextState provides KV state store operations for a run.
type RunContextState struct {
	Get    func(key string) (any, error)
	Set    func(key string, value any) error
	Delete func(key string) error
	List   func() ([]map[string]any, error)
}

// RunContext is the context object passed to a job's run handler.
type RunContext struct {
	// RunID is the unique identifier for this run.
	RunID string
	// Attempt is the current attempt number (1-based).
	Attempt int
	// Ctx is the context for the run, supporting cancellation.
	Ctx context.Context
	// Logger provides structured logging for the run.
	Logger *slog.Logger
	// Checkpoint saves intermediate state for crash recovery.
	Checkpoint func(state map[string]any) error
	// ReportProgress reports execution progress (0.0 to 1.0).
	ReportProgress func(percent float64, message ...string) error
	// Heartbeat signals that the run is still alive.
	Heartbeat func() error
	// ReportUsage reports LLM token usage and cost.
	ReportUsage func(usage UsageReport) error
	// LogToolCall logs a tool invocation.
	LogToolCall func(toolCall ToolCallReport) error
	// SaveOutput saves a named output value.
	SaveOutput func(key string, value map[string]any, schema ...map[string]any) error
	// State provides KV state store operations.
	State *RunContextState
	// StreamChunk sends a streaming chunk.
	StreamChunk func(chunk string, opts ...StreamChunkOption) error
	// WaitForEvent pauses until an external event is received.
	WaitForEvent func(eventKey string, opts ...WaitForEventOption) (map[string]any, error)
	// Spawn creates a child run.
	Spawn func(opts SpawnOptions) (map[string]any, error)
	// Continue continues the run with an optional payload.
	Continue func(payload ...map[string]any) (map[string]any, error)
	// Annotate adds annotations to the run.
	Annotate func(annotations map[string]string) error
	// Complete marks the run as complete.
	Complete func(result ...map[string]any) error
	// Fail marks the run as failed.
	Fail func(errMsg string) error
}

// UsageReport captures LLM token usage and cost.
type UsageReport struct {
	Provider         string
	Model            string
	PromptTokens     *int
	CompletionTokens *int
	TotalTokens      *int
	CostMicrousd     *int
}

// ToolCallReport captures a tool invocation.
type ToolCallReport struct {
	ToolName   string
	Input      map[string]any
	Output     map[string]any
	DurationMs *int
	Status     string
}

// SpawnOptions configures a child run.
type SpawnOptions struct {
	JobSlug   string
	ProjectID string
	Payload   map[string]any
	Priority  *int
}

// StreamChunkOption configures stream chunk options.
type StreamChunkOption struct {
	StreamID string
	Done     bool
}

// WaitForEventOption configures wait for event options.
type WaitForEventOption struct {
	TimeoutSecs *int
	NotifyUrl   string
}

File: packages/ruby-sdk/lib/strait/authoring/run_context.rb
# frozen_string_literal: true

module Strait
  module Authoring
    # KV state store operations for a run.
    class RunContextState
      attr_reader :get, :set, :delete, :list

      def initialize(get:, set:, delete:, list:)
        @get = get
        @set = set
        @delete = delete
        @list = list
      end
    end

    # Context object passed to a job's run handler.
    class RunContext
      attr_accessor :run_id, :attempt, :checkpoint, :report_progress, :heartbeat,
                    :report_usage, :log_tool_call, :save_output, :state,
                    :stream_chunk, :wait_for_event, :spawn, :continue_run,
                    :annotate, :complete, :fail

      def initialize(
        run_id:,
        attempt: 1,
        checkpoint: nil,
        report_progress: nil,
        heartbeat: nil,
        report_usage: nil,
        log_tool_call: nil,
        save_output: nil,
        state: nil,
        stream_chunk: nil,
        wait_for_event: nil,
        spawn: nil,
        continue_run: nil,
        annotate: nil,
        complete: nil,
        fail: nil
      )
        @run_id = run_id
        @attempt = attempt
        @checkpoint = checkpoint
        @report_progress = report_progress
        @heartbeat = heartbeat
        @report_usage = report_usage
        @log_tool_call = log_tool_call
        @save_output = save_output
        @state = state
        @stream_chunk = stream_chunk
        @wait_for_event = wait_for_event
        @spawn = spawn
        @continue_run = continue_run
        @annotate = annotate
        @complete = complete
        @fail = fail
      end
    end
  end
end
